### PR TITLE
[PFCWD + Asym PFC]: Allow PFCWD to detect PFC storm on all priorities when Asym PFC is enabled

### PR DESCRIPTION
**Pull request by volodymyrsamotiy**
Signed-off-by: Volodymyr Samotiy <volodymyrs@mellanox.com>

**What I did**
Add possibility for PFCWD  to detect PFC storm on all priorities when asymmetric PFC is enabled.

According to requirements PFCWD should detect PFC storm on the lossy priorities for the port on which asymmetric PFC is enabled.
(when asymmetric PFC is enabled on some port that means such port is enabled for receiving PFC frames on all priorities)

**Why I did it**
To fix issue https://github.com/Azure/sonic-swss/issues/1128.

**How I verified it**
1. Enable asymmetric PFC on test port.
```
pfc config asymmetric on <port>
```
2. Enable PFCWD on the test port.
```
pfcwd start --action drop --restoration-time 3000 <port> 1000
pfcwd interval 500
```
3. Run ```pause``` frames to the test port for all priorities.
4. Verify that PFCWD detected storm for all priorities.
```
pfcwd show stats
```

**Details if related**
N/A

### CODE REVIEW

**Comment by liat-grozovik** [comment]
retest vs please

**Comment by volodymyrsamotiy** [comment]
retest this please

**Comment by liat-grozovik** [comment]
retest vs please

**Comment by volodymyrsamotiy** [comment]
retest this please

**Comment by liat-grozovik** [comment]
retest this please

**Comment by volodymyrsamotiy** [comment]
retest this please

**Comment by volodymyrsamotiy** [comment]
retest this please

File: orchagent/portsorch.cpp
Comment (liat-grozovik) [inline orchagent/portsorch.cpp:891]: not sure i understand this part.
you are using the tx pfc bitmask for setting flow control and the message in the log is still pfc bitmask which i can see later on you have one for rx and tx.

File: orchagent/portsorch.cpp
Comment (liat-grozovik) [inline orchagent/portsorch.cpp:965]: logic has changed, can you elaborate it in the code?

File: orchagent/portsorch.cpp
Comment (liat-grozovik) [inline orchagent/portsorch.cpp:977]: IMO you should remove 'the' after the 'for'

File: orchagent/qosorch.cpp
Comment (liat-grozovik) [inline orchagent/qosorch.cpp:1332]: the pfc_enable is on/off? if this is the case all other places are mask. can you please elaborate on that?

**Comment by liat-grozovik** [comment]
retest this please

**Comment by liat-grozovik** [comment]
@volodymyrsamotiy  please handle conflicts

File: orchagent/portsorch.cpp
Comment (volodymyrsamotiy) [reply orchagent/portsorch.cpp:957]: This logic is explained in the "Asymmentric PFC" HLD in the "2.2.1 PortsOrch" section
https://github.com/Azure/SONiC/wiki/Asymmetric-PFC-High-Level-Design#221-portsorch

File: orchagent/portsorch.cpp
Comment (volodymyrsamotiy) [reply orchagent/portsorch.cpp:1031]: Logic is exactly the same but code was refactored according to changes for set/getPortPfc API.

File: orchagent/qosorch.cpp
Comment (volodymyrsamotiy) [reply orchagent/qosorch.cpp:1336]: Yes, it is according to requirements, "Asym PFC" is "on"/"off" port attribute for the user but SAI requires PFC value to be as bit vector.
https://github.com/Azure/SONiC/wiki/Asymmetric-PFC-High-Level-Design#11-functional-requirements

File: orchagent/portsorch.cpp
Comment (volodymyrsamotiy) [reply orchagent/portsorch.cpp]: changed

**Comment by volodymyrsamotiy** [comment]
> @volodymyrsamotiy please handle conflicts

handled

**Review by liat-grozovik — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@neethajohn  kindly reminder. this PR should go to 202012 as well.

**Comment by dprital** [comment]
@neethajohn Kindly Reminder, we are waiting for the merge of this PR. Thanks.

**Comment by dprital** [comment]
@neethajohn  - Who can merge this PR ? Thanks.

File: orchagent/pfcwdorch.cpp
Comment (neethajohn) [inline orchagent/pfcwdorch.cpp:494]: How do you handle the case where pfcwd is already started with asym pfc disabled for a port and later asym pfc is enabled? In that case, for the pfcwd to act on all priorities, we need to restart it.

File: orchagent/pfcwdorch.cpp
Comment (neethajohn) [inline orchagent/pfcwdorch.cpp:479]: We need some comments in here since now 'losslessTc' var will contain all priorities in the case of asym pfc

File: orchagent/pfcactionhandler.cpp
Comment (neethajohn) [inline orchagent/pfcactionhandler.cpp:396]: Why do we need to modify Tx Mask?  In any case, its a noop if pfcwd is triggered on a lossy priority because the txMask only contains priorities 3 and 4

File: orchagent/pfcactionhandler.cpp
Comment (neethajohn) [inline orchagent/pfcactionhandler.cpp:417]: We shouldn't be modifying Tx Mask since it contains only priorities 3 and 4. We will setting tx mask to incorrect values on pfcwd storm restore if pfcwd was triggered on a lossy priority

**Review by neethajohn — COMMENTED** [review]
Please add unit tests

File: orchagent/pfcwdorch.cpp
Comment (volodymyrsamotiy) [reply orchagent/pfcwdorch.cpp:479]: Added below where ```losslessTc``` is declared.

File: orchagent/portsorch.cpp
Comment (volodymyrsamotiy) [reply orchagent/portsorch.cpp]: Fixed.

File: orchagent/pfcactionhandler.cpp
Comment (volodymyrsamotiy) [reply orchagent/pfcactionhandler.cpp]: Agree, we don' need it. Fixed to modify only Rx mask.

File: orchagent/pfcactionhandler.cpp
Comment (volodymyrsamotiy) [reply orchagent/pfcactionhandler.cpp]: Agree, we don' need it. Fixed to modify only Rx mask.

File: orchagent/pfcwdorch.cpp
Comment (volodymyrsamotiy) [reply orchagent/pfcwdorch.cpp:496]: Yes, we need to restart it. Original plan was to leave it under user responsibility. So once PFC configuration is changed then PFCWD should be restarted in order to apply the changes for PFCWD. BTW, I am not sure if it is a good idea to restart PFCWD automatically because it might affect some traffic flows. And user knows when to restart PFCWD in order to avoid or minimize traffic impact. Also usually both PFCWD and Asymmetric PFC are configured once in JSON config file and in such case all should be working without any restarts.

File: orchagent/port.h
Comment (bratashX) [inline orchagent/port.h:142]: We have two variable for save _port priority flow control mode_: `m_pfc_asym` and  `pfc_mode`  in  `struct PfcInfo`
 Maybe we can remove `m_pfc_asym` ?

File: orchagent/pfcwdorch.cpp
Comment (neethajohn) [reply orchagent/pfcwdorch.cpp:496]: I agree it is not a good idea to restart pfcwd automatically. but we do have a cli option ' pfc config asymmetric' to enable/disable asymmetric pfc. So we need to add in some comments so that user is aware that pfcwd will see these changes only after a restart

**Review by neethajohn — COMMENTED** [review]
Please add unit tests

**Comment by liat-grozovik** [comment]
> Please add unit tests

@neethajohn done. Please review

File: tests/test_pfcwd.py
Comment (neethajohn) [inline tests/test_pfcwd.py:167]: Add cleanup part of this testcase

Comment (volodymyrsamotiy) [reply tests/test_pfcwd.py:166]: Done

**Comment by liat-grozovik** [comment]
@neethajohn  kindly reminder to approve this PR. once approved we will issue a new one against 202012 as it is not cleanly cherry picked.

File: orchagent/pfcactionhandler.cpp
Comment (neethajohn) [inline orchagent/pfcactionhandler.cpp:424]: Need to update pfcTxMask as well in combined mode and in asym mode, update pfcTxMask if the queue is one specified in 'pfc_enable' fields of PORT_QOS_MAP

**Review by neethajohn — CHANGES_REQUESTED** [review]

File: orchagent/pfcactionhandler.cpp
Comment (neethajohn) [reply orchagent/pfcactionhandler.cpp]: Thinking about this more, we need to set the txmask but need to do it only for queues set in 'pfc_enable' field

**Comment by neethajohn** [comment]
@volodymyrsamotiy , I think these unit tests are not sufficient to ensure that pfcwd feature is working correctly for asym mode and in combined mode. I am concerned that there might be scenarios we miss and end up causing regression. Can we add unit tests for the following flow in both combined and asym mode?
1. start pfcwd
2. verify the correct set of counters are added in flex counter for polling
3. in asic db, verify that the pfc mask is getting disabled for the right queue/queues in storm
4. stop pfcwd, in asic db, verify that pfc mask is enabled for the right queue

File: orchagent/pfcactionhandler.cpp
Comment (volodymyrsamotiy) [reply orchagent/pfcactionhandler.cpp]: Ok, I will change it back and verify

**Comment by volodymyrsamotiy** [comment]
@neethajohn, I agree that unit test checks only simple logic but that is because we are pretty limited on virtual switch. Currently we don't have any VS test for PFCWD and it was never checked on VS so maybe even some additional VS infra will required. A while ago I already tried it on virtual switch and PFCWD logic is not fully functional on it, so test checks everything we currently can check.
In any case I will think about it more and check what else we can do.
Regarding "sonic-mgmt" test, I will first fix latest comments for the "tx mask", then verify and let you know.

**Comment by liat-grozovik** [comment]
@neethajohn  do you have additional comments or you can sign off?

**Comment by neethajohn** [comment]
> @neethajohn do you have additional comments or you can sign off?

@liat-grozovik , I don't see my comments regarding vs tests addressed. In the last review I did, there were issues in the logic that could have been caught by the vs test flow I had suggested. Also I didn't hear back on whether the current pfcwd and asym pfc testcases in sonic-mgmt are passing with this change.

**Comment by liat-grozovik** [comment]
@neethajohn any further comments or we can close this one?

**Comment by liat-grozovik** [comment]
@neethajohn  following an offline discussion regarding tests, we agree that what we have the tests based on the existing infra and if extending the infra is required for pfc wd it should be discussed regardless of this PR. There is a sonic-mgmt test to cover this issue.
As of that, could you please review and let us know if there are any additional outstanding issues that this PR should address?

**Comment by liat-grozovik** [comment]
@neethajohn please note comments were addressed. please review and signoff.

**Comment by dprital** [comment]
@neethajohn - Can you please approve this PR ?

**Comment by dprital** [comment]
@neethajohn  - as discussed yesterday, please review. Thanks.

**Comment by liat-grozovik** [comment]
@volodymyrsamotiy  please handle conflicts and we will have it merged.

**Comment by liat-grozovik** [comment]
@volodymyrsamotiy kindly reminder to handle conflicts so I can merge

**Comment by liat-grozovik** [comment]
The issue was reported many many releases a go.
due to very low attention from the community and probably not in interest of others, i am closing this PR.
if at any time in the future the fix will be needed, we will need community attention. till then the issues is open

**Pull request closed without merging**
